cmd/tools/faultinjection/main.go

```diff
@@ -0,0 +1,230 @@
+
+	for _, store := range stores() {
+		callWithFile(generateFaultInjectionStore, store, *outPathFlag, licenseText)
+	}
+
+}
+
+func stores() []reflect.Type {
+	factoryT := reflect.TypeOf((*persistence.DataStoreFactory)(nil)).Elem()
+	var storeTs []reflect.Type
+	for i := 0; i < factoryT.NumMethod(); i++ {
+		ctor := factoryT.Method(i)
+		if !strings.HasPrefix(ctor.Name, "New") {
+			continue
+		}
+		if ctor.Type.NumOut() == 0 {
+			continue
+		}
+		storeT := ctor.Type.Out(0)
+		storeTs = append(storeTs, storeT)
+	}
+	return storeTs
+}
+
+func callWithFile(generator func(io.Writer, reflect.Type), store reflect.Type, outPath string, licenseText string) {
+	filename := path.Join(outPath, camelCaseToSnakeCase(store.Name())+"_gen.go")
+	w, err := os.Create(filename)
+	panicIfErr(err)
+	defer func() {
+		panicIfErr(w.Close())
+	}()
+	_, err = fmt.Fprintf(w, "%s\n// Code generated by cmd/tools/faultinjection. DO NOT EDIT.\n", licenseText)
+	panicIfErr(err)
+	generator(w, store)
+}
+
+func generateFaultInjectionStore(w io.Writer, store reflect.Type) {
+	writeStoreHeader(w, store, `
+package faultinjection
+
+import (
+	"context"
+{{.ExtraImports}}
+	"go.temporal.io/server/common/persistence"
+)
+
+type (
+	faultInjection{{.StoreName}} struct {
+		baseStore persistence.{{.StoreName}}
+		generator faultGenerator
+	}
+)
+
+func newFaultInjection{{.StoreName}}(
+	baseStore persistence.{{.StoreName}},
+	generator faultGenerator,
+) *faultInjection{{.StoreName}} {
+	return &faultInjection{{.StoreName}}{
+		baseStore: baseStore,
+		generator: generator,
+	}
+}
+`)
+
+	writeStoreMethods(w, store, `
+func (c *faultInjection{{.StoreName}}) {{.MethodName}}(
+{{.InParams}}) {{.OutParams}} {
+	return inject{{.NumOutParams}}(c.generator.generate("{{.MethodName}}"), func() {{.OutParams}} {
+		return c.baseStore.{{.MethodName}}({{.InVars}})
+	})
+}
+`, `
+func (c *faultInjection{{.StoreName}}) {{.MethodName}}(
+{{.InParams}}) {{.OutParams}} {
+	{{if ne .OutParams ""}}return {{end}}c.baseStore.{{.MethodName}}({{.InVars}})
+}
+`)
+}
+
+func writeStoreHeader(w io.Writer, store reflect.Type, tmpl string) {
+	panicIfErr(template.Must(template.New("code").Parse(tmpl)).Execute(w, map[string]string{
+		"StoreName":    store.Name(),
+		"ExtraImports": extraImports[store.Name()],
+	}))
+}
+
+func writeStoreMethods(w io.Writer, store reflect.Type, tmpl string, passThroughTmpl string) {
+	for i := 0; i < store.NumMethod(); i++ {
+		writeStoreMethod(w, store, store.Method(i), tmpl, passThroughTmpl)
+	}
+}
+
+func writeStoreMethod(w io.Writer, store reflect.Type, m reflect.Method, tmpl string, passThroughTmpl string) {
+
+	mT := m.Type
+	inParams := ""
+	inVars := ""
+	for i := 0; i < mT.NumIn(); i++ {
+		pName := fmt.Sprintf("p%d", i)
+		if mT.In(i).String() == "context.Context" {
+			pName = "ctx"
+		} else if strings.HasSuffix(mT.In(i).String(), "Request") {
+			pName = "request"
+		}
+		inParams += fmt.Sprintf("\t %s %s, \n", pName, mT.In(i).String())
```

Comment:
you could probably do the whitespace with some template code. but fine either way